### PR TITLE
154 improve ncbi genomic region generator

### PR DESCRIPTION
**Pull request by jonas-hag**
### Summary

  This PR improves NCBI genomic region generation by adding robust assembly resolution modes, stricter validation/error handling, and expanded test coverage (including edge cases for missing transcript/annotation
  metadata).

  ### What changed

  1. NCBI FTP loader enhancements (`_ftp_loader.py`)

  - Added two supported retrieval modes:
      - Taxon/species mode (taxon, species, annotation_release, optional assembly_source)
      - Direct assembly mode (refseq_assembly_accession + assembly_name)
  - Added assembly_source support (auto, annotation_releases, latest_assembly_versions, reference) with taxon/source compatibility validation.
  - Added strong config validation and explicit ConfigurationErrors for invalid parameter combinations.
  - Improved assembly metadata detection with fallback from README_* to *_assembly_report.txt.
  - Added annotation release metadata extraction from NCBI README where available; defaults to no_annotation_info and emits warning if unavailable.
  - Added support for direct genomes/all/... resolution using RefSeq accession parsing.

  2. Genomic region generator robustness (`_genomic_region_generator.py`)

  - Region extraction now raises explicit errors when requested annotation types are missing.
  - Added explicit errors when intron/UTR/exon-exon-junction computation yields empty output.
  - Transcript-count handling is now aware of required gene_ids:
      - Attempts to compute counts for all relevant gene_ids
      - Falls back with warning (instead of failing) when transcript info is partially/unavailable
  - FASTA header construction updated so number_total_transcripts is included only when validly computed.

  3. Config updates

  - Updated NCBI config docs/examples to describe:
      - Mode A (taxon/species)
      - Mode B (direct assembly accession + assembly name)
      - assembly_source behavior and constraints

  4. Test suite expansion

  - Added new test module with expected FASTA header values across human/mouse/bacteria scenarios in `tests/test_sequence_generator.py`.
  - Extended FTP loader tests:
      - mode validation
      - source/taxon compatibility
      - current-release resolution behavior
      - metadata fallback behavior
  - Extended region generator tests:
      - pass/warn/error expectations across species/data completeness cases
      - FASTA header field correctness assertions
  - Added large NCBI test fixtures (GTF/FASTA/FAI) for mouse and bacteria edge-case coverage.

related to #154

### CODE REVIEW

**Comment by jonas-hag** [comment]
@codex review

File: data/configs/genomic_region_generator_ncbi.yaml [hunk 1]
```diff
@@ -8,9 +8,15 @@ dir_output: output_genomic_region_generator_ncbi # name of the directory where t
...
-  annotation_release: 110 # required: release number of annotation e.g. '109' or '109.20211119'  or 'current' to use most recent annotation release. Check out release numbers for NCBI at ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/
+  # Mode A (taxon/species based): set taxon/species/assembly_source/annotation_release
+  taxon: vertebrate_mammalian # required in Mode A; supported taxa are: archaea, bacteria, fungi, invertebrate, metagenomes, plants, protozoa, unknown, vertebrate_mammalian, vertebrate_other, viral
+  species: Homo_sapiens # required in Mode A; species name in NCBI download format
+  assembly_source: auto # optional in Mode A: auto, annotation_releases, latest_assembly_versions, reference
+  annotation_release: 110 # required in Mode A; use release number only with assembly_source=annotation_releases; otherwise set to 'current'
+
+  # Mode B (direct assembly): set both fields below and set Mode A fields above to None
+  refseq_assembly_accession: # GCF_000001405.38
+  assembly_name: # GRCh38.p12
```
Comment (lisa-sousa) [inline data/configs/genomic_region_generator_ncbi.yaml:27]: Does this ymal file work as example case when both modes are parametrized? Which mode will be used?

File: data/configs/genomic_region_generator_ncbi.yaml [hunk 2]
```diff
@@ -8,9 +8,15 @@ dir_output: output_genomic_region_generator_ncbi # name of the directory where t
...
 source_params:
-  taxon: vertebrate_mammalian # required: taxon of the species, valid taxa are: archaea, bacteria, fungi, invertebrate, mitochondrion, plant, plasmid, plastid, protozoa, vertebrate_mammalian, vertebrate_other, viral
-  species: Homo_sapiens # required: species name in NCBI download format, e.g. 'Homo_sapiens' for human; see https://ftp.ncbi.nlm.nih.gov/genomes/refseq/ for available species name
-  annotation_release: 110 # required: release number of annotation e.g. '109' or '109.20211119'  or 'current' to use most recent annotation release. Check out release numbers for NCBI at ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/
+  # Mode A (taxon/species based): set taxon/species/assembly_source/annotation_release
+  taxon: vertebrate_mammalian # required in Mode A; supported taxa are: archaea, bacteria, fungi, invertebrate, metagenomes, plants, protozoa, unknown, vertebrate_mammalian, vertebrate_other, viral
+  species: Homo_sapiens # required in Mode A; species name in NCBI download format
+  assembly_source: auto # optional in Mode A: auto, annotation_releases, latest_assembly_versions, reference
```
Comment (lisa-sousa) [inline data/configs/genomic_region_generator_ncbi.yaml]: are: annotation_releases, latest_assembly_versions, reference
examples of what this parameter can contain? If yes, can you specify and also explain what autp means?

File: oligo_designer_toolsuite/pipelines/_genomic_region_generator.py [hunk 3]
```diff
@@ -82,6 +83,9 @@ def load_annotations(
                 taxon=source_params["taxon"],
                 species=source_params["species"],
                 annotation_release=source_params["annotation_release"],
+                assembly_source=source_params["assembly_source"],
+                refseq_assembly_accession=source_params["refseq_assembly_accession"],
```
Comment (lisa-sousa) [inline oligo_designer_toolsuite/pipelines/_genomic_region_generator.py:95]: Can those be None? If mode A is used

File: oligo_designer_toolsuite/sequence_generator/_ftp_loader.py [hunk 4]
```diff
@@ -228,21 +230,61 @@ class FtpLoaderNCBI(BaseFtpLoader):
...
+        "plant": {"annotation_releases", "latest_assembly_versions", "reference"},
+        "plants": {"annotation_releases", "latest_assembly_versions", "reference"},
+        "protozoa": {"latest_assembly_versions", "reference"},
+        "unknown": {"latest_assembly_versions"},
+        "vertebrate_mammalian": {"annotation_releases", "latest_assembly_versions", "reference"},
+        "vertebrate_other": {"annotation_releases", "latest_assembly_versions", "reference"},
+        "viral": {"latest_assembly_versions"},
+    }
+    UNSUPPORTED_TAXA: set[str] = {"mitochondrion", "plasmids", "plastid"}
+    ALLOWED_ASSEMBLY_SOURCES: set[str] = {
```
Comment (lisa-sousa) [inline oligo_designer_toolsuite/sequence_generator/_ftp_loader.py:269]: can you specify as comment what those options mean?

File: oligo_designer_toolsuite/sequence_generator/_ftp_loader.py [hunk 5]
```diff
@@ -289,44 +333,265 @@ def _get_params(self, file_type: _TYPES_FILE) -> tuple[str, str, str]:
...
+                )
+            if self.assembly_source != "auto":
+                raise ConfigurationError(
+                    "assembly_source cannot be set when using direct assembly mode "
+                    "(refseq_assembly_accession/assembly_name)."
+                )
+            self.annotation_release = "current"
+            return True
+
+        if self.taxon is None or self.species is None or self.annotation_release is None:
```
Comment (lisa-sousa) [inline oligo_designer_toolsuite/sequence_generator/_ftp_loader.py]: it doesn not return True if only taxon_mode_params are provided, no?

File: oligo_designer_toolsuite/sequence_generator/_ftp_loader.py [hunk 6]
```diff
@@ -289,44 +333,265 @@ def _get_params(self, file_type: _TYPES_FILE) -> tuple[str, str, str]:
...
+            self.annotation_release = entries[0]
+        return f"{base_directory}{self.annotation_release}/"
 
-        ftp_directory = ftp_directory + f"{self.annotation_release}/"
+    def _resolve_directory_from_direct_assembly(self) -> str:
+        if not self.assembly_accession or not self.assembly_name:
+            raise ConfigurationError(
+                "Direct assembly mode requires both 'refseq_assembly_accession' and 'assembly_name'."
+            )
+        match = re.match(r"^(GCF)_(\d+)\.(\d+)$", self.assembly_accession)
```
Comment (lisa-sousa) [inline oligo_designer_toolsuite/sequence_generator/_ftp_loader.py:499]: are we matchign always to GCF?

File: oligo_designer_toolsuite/sequence_generator/_ftp_loader.py [hunk 7]
```diff
@@ -289,44 +333,265 @@ def _get_params(self, file_type: _TYPES_FILE) -> tuple[str, str, str]:
...
+        if match is None:
+            raise ConfigurationError(
+                f"Invalid RefSeq assembly accession '{self.assembly_accession}'. "
+                "Expected format like 'GCF_000001405.38'."
+            )
+        accession_prefix = match.group(1)
+        accession_number = match.group(2)
+        if len(accession_number) != 9:
```
Comment (lisa-sousa) [inline oligo_designer_toolsuite/sequence_generator/_ftp_loader.py:507]: is the length of 9 fixed?

File: oligo_designer_toolsuite/sequence_generator/_genomic_region_generator.py [hunk 8]
```diff
@@ -425,6 +421,17 @@ def get_sequence_exon(self, collapse_duplicated_regions: bool = True) -> str:
         if collapse_duplicated_regions:
             annotation = self._collapse_duplicated_regions(annotation)
 
+        # add transcript counts for each gene
```
Comment (lisa-sousa) [inline oligo_designer_toolsuite/sequence_generator/_genomic_region_generator.py:424]: this section isrepeated for each region type -> can we move ut to a function?

File: oligo_designer_toolsuite/sequence_generator/_genomic_region_generator.py [hunk 9]
```diff
@@ -1221,34 +1277,54 @@ def __init__(
...
-            taxon = "vertebrate_mammalian"
-            warnings.warn(f"No taxon defined. Using default taxon {taxon}!")
+        has_direct_mode = refseq_assembly_accession is not None or assembly_name is not None
 
-        if species is None:
-            species = "Homo_sapiens"
-            warnings.warn(f"No species defined. Using default species {species}!")
+        if has_direct_mode and assembly_source is None:
+            assembly_source = "auto"
+        elif not has_direct_mode:
```
Comment (lisa-sousa) [inline oligo_designer_toolsuite/sequence_generator/_genomic_region_generator.py]: I thibk having a default species doesnt make sense. i should throw a warning instead

**Review by lisa-sousa — CHANGES_REQUESTED** [review]
Generally looks good. A few main points, additionally to the comments:
- the yaml file should work as provided. Currently it has mode A and mode B active. Does this work?
- We should not se any defaults for species/taxa
- The transcript section is repetitive in each region generator function. Can we have this in a seperate function?
- document modes and their parameters a bit more in the yaml (or later in the pydantic model)

File: oligo_designer_toolsuite/sequence_generator/_ftp_loader.py (hunk 6 above)
Comment (jonas-hag) [reply oligo_designer_toolsuite/sequence_generator/_ftp_loader.py:499]: yes, I decided to always use GCF as all the examples I've seen so far are GCF and in general they are a more reliable source than GCA as they are curated by NCBI

File: oligo_designer_toolsuite/sequence_generator/_ftp_loader.py (hunk 7 above)
Comment (jonas-hag) [reply oligo_designer_toolsuite/sequence_generator/_ftp_loader.py:507]: yes, this is the format currently used by NCBI

File: oligo_designer_toolsuite/pipelines/_genomic_region_generator.py (hunk 3 above)
Comment (jonas-hag) [reply oligo_designer_toolsuite/pipelines/_genomic_region_generator.py:95]: yes, it was parsed but I agree that it's not optimal. I now switched it to a better structure

File: oligo_designer_toolsuite/sequence_generator/_ftp_loader.py (hunk 5 above)
Comment (jonas-hag) [reply oligo_designer_toolsuite/sequence_generator/_ftp_loader.py]: updated the function with new parameter to directly determine the mode; should be obsolete

File: data/configs/genomic_region_generator_ncbi.yaml (hunk 1 above)
Comment (jonas-hag) [reply data/configs/genomic_region_generator_ncbi.yaml:27]: would throw an error. But I have no changed it so that you need to specify the mode

File: oligo_designer_toolsuite/sequence_generator/_genomic_region_generator.py (hunk 9 above)
Comment (jonas-hag) [reply oligo_designer_toolsuite/sequence_generator/_genomic_region_generator.py]: changed taxon & species to errors

File: oligo_designer_toolsuite/sequence_generator/_genomic_region_generator.py (hunk 8 above)
Comment (jonas-hag) [reply oligo_designer_toolsuite/sequence_generator/_genomic_region_generator.py:424]: moved adding transcript counts into function but kept BED headers separately as they differ a bit for UTR and exon-exon junctions

**Comment by jonas-hag** [comment]
thanks for the review; all comments should be addressed now

File: oligo_designer_toolsuite/sequence_generator/_ftp_loader.py [hunk 10]
```diff
@@ -289,44 +343,268 @@ def _get_params(self, file_type: _TYPES_FILE) -> tuple[str, str, str]:
...
+                raise ConfigurationError(
+                    "In mode='assembly', taxon/species/annotation_release must not be provided."
+                )
+            if self.assembly_source != "auto":
+                raise ConfigurationError(
+                    "In mode='assembly', assembly_source cannot be set and must remain 'auto'."
+                )
+            self.annotation_release = "unknown"
+            return True
+
```
Comment (lisa-sousa) [inline oligo_designer_toolsuite/sequence_generator/_ftp_loader.py]: for readability it would be ncie to have an "elif" statement here such that it's easier to get that we are here ocnsidering mode=species

**Review by lisa-sousa — CHANGES_REQUESTED** [review]
Just one small comment, otherwise looks good now!

File: oligo_designer_toolsuite/sequence_generator/_ftp_loader.py (hunk 10 above)
Comment (jonas-hag) [reply oligo_designer_toolsuite/sequence_generator/_ftp_loader.py]: thanks; I've refactored this part to make it clearer; we didn't really need the bool return value

**Review by lisa-sousa — COMMENTED** [review]
Looks all good now

**Review by lisa-sousa — APPROVED** [review]
Looks all good now
